include/openssl_shim.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <vector>
+#include <system_error>
+#include <dlfcn.h>
+
+#ifdef _WIN32
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#include <bcrypt.h>
+#else
+#ifdef USE_OPENSSL
+#include <openssl/hmac.h>
+#else
+#include <gnutls/gnutls.h>
+#include <gnutls/crypto.h>
+#endif
+#endif
+
+namespace azure {  namespace storage_lite {
+
+  extern void* dl_handle;
+  extern std::string dl_error_;
+  extern unsigned long ossl_ver;
+
+  inline void clear_dlerror() {
+    dl_error_ = std::string("");
+    dlerror();
+  }
+  
+  inline void set_dlerror() {
+    char *errmsg = dlerror();
+    if(errmsg) {
+      dl_error_.empty() ? (dl_error_ = errmsg) : (dl_error_ += std::string("\n") + errmsg);
+    }
+  }
+
+  void ossl_shim_init(void);
+  void *get_dlopen_handle(const std::string& name, const std::string& version); 
+  void *get_dlopen_handle(const std::string& name); 
+
+#define BIND_SYMBOL(H, X, Y, Z)  \
+  do {                           \
+    clear_dlerror();             \
+    X = Z dlsym(H, Y);           \
+    if (!X) {                    \
+      set_dlerror();             \
+      throw std::system_error(ECANCELED, std::generic_category(), dl_error_); \
+    }                            \
+  } while (false)
```

Comment:
What's the reason for this `do...while(false)` here?

Reply:
This is code that @Santhosh-K-M copied from TileDB. The reasoning is to keep multi-statement code in a `#define` contained in a block for instances where you think it is a single statement. For example : `if (x) bind(x);` would get incorrectly translated to `if (x) stmt 1; stmt2` whereas with the do-while it will get correctly translated to `if (x) do {stm11, stmt2} while (false)`.
You will see this pattern used all over the C runtimes for example. See https://stackoverflow.com/questions/154136/why-use-apparently-meaningless-do-while-and-if-else-statements-in-macros

Reply:
Ah - got it. Thanks for the tip!